Add “Using with React” doc (#577)
manifoldco/ui: docs/docs/advanced/react-events.md

---
title: Using with React
path: /advanced/using-with-react
---

# Using with React

Since UI are standard web components, that means React support is fully-built in since web
components are now a fully-supported browser standard. However, if you’re new to web components,
there can be some gotchas. Here are the tough parts we’ve found while implementing UI in a React

> environment.

Reviewer: Don't accept the suggestion in GH because the diff is wrong.

Just some more concise wording that you can take or leave.

```suggestion
Since Manifold UI components are standard Web Components, React is fully-supported. However, 
if you’re new to Web Components, there can be some gotchas. Here are some of the challenges 
we've experienced while implementing Manifold UI with React.
```

## Rule 1: Events must be handled with listeners

Handling web component events in React is arguably the hardest part. In React, you may be used to
handy `onClick={() => }` props for easy callbacks. Unfortunately that’s not yet possible with web
components. Listeners are required. In the React projects we’ve worked in, we’ve found [React

> Hooks][hooks] (specifically `useState` and `useRef`) to be the simplest form of handling this.

Reviewer: ```suggestion
Hooks][hooks] (specifically `useState` and `useRef`) helpful for managing event listeners on Web Components.
```

### Example

Say you’re integrating the [Plan Selector][plan-selector] component in React, and you want to be

> notified every time a user changes their plan to pass to the [provision-button]. This is what that

Reviewer: ```suggestion
notified when the plan changes in order to to pass the selected plan to the [provision-button]. This is what that
```

1. Within the effect, we’re binding the `manifold-planSelector-load` and
   `manifold-planSelector-change` events to an `update()` function (we know what they are because we
   looked at the [Plan Selector doc][plan-selector]).
1. Whenever `update()` fires, we’re telling React to re-render with the new props for
   `<manifold-data-provision-button>`.

Lastly, the `if (planSelectorRef.current)(` check may be a little confusing. It’s here because if
using TypeScript you’ll find you’ll need to handle the case `ref.current` isn’t a DOM element. But
as long as `<manifold-plan-selector>` is present on the page (i.e. isn’t conditionally rendered or
not), we should have no problems here.

### How do I find what events are possible?

Most Manifold UI components emit several events. To see what each component emits, please reference
the **Events** section of that component’s documentation.

## Rule 2: Every prop is a string

Another reason consuming web component events can be difficult within React is because whereas React
has no problem passing any data type (`object`, `array`, `number`, etc.) as a prop, **web components

> can only accept strings** as they’re HTML:

Reviewer: Don't accept suggestion in GH. Diff is wrong

```suggestion
Another reason consuming web component events can be difficult within React is because **web components can only accept strings as props** because they’re just regular HTML elements:
```
